src/conn/conn_compact.c

```diff
@@ -0,0 +1,196 @@
+    WT_SESSION *wt_session;
+    WT_SESSION_IMPL *session;
+    int exact;
+    const char *key;
+    bool full_iteration;
+
+    session = arg;
+    conn = S2C(session);
+    wt_session = (WT_SESSION *)session;
+    cursor = NULL;
+    key = NULL;
+    exact = 0;
+    full_iteration = false;
+
+    for (;;) {
+
+        /* When the entire metadata file has been parsed, take a break. */
+        if (full_iteration) {
+            full_iteration = false;
+            /*
+             * TODO: We may have processed just a few files, might as well sleep to give the system
+             * a break.
+             */
+            __wt_sleep(10, 0);
+        }
+
+        /* Only an explicit call to the compact API should signal the thread to wake up. */
+        if (!conn->background_compact.running) {
+            /*
+             * TODO: Depending on the previous state, we may not want to clear out the last key
+             * used. This could be useful if the server was paused to be resumed later.
+             */
+            key = NULL;
+            __wt_cond_wait(session, conn->background_compact.cond, 0, NULL);
+        }
+
+        /*
+         * Now the server has been requested to run, check if it is supposed to, we could be
+         * quitting or reconfiguring the connection too.
+         */
+        if (!__compact_server_run_chk(session))
+            break;
+
+        /* Open a metadata cursor. */
+        WT_ERR(__wt_metadata_cursor(session, &cursor));
+
+        /*
+         * If we are in the middle of an iteration, position the cursor on the previously selected
+         * key.
+         */
+        if (key != NULL) {
+            cursor->set_key(cursor, key);
+            WT_ERR(cursor->search_near(cursor, &exact));
+        }
+
+        /*
+         * If no key has been set yet, move to the first record. If the previously used key could
+         * not be found, make sure not to go backwards.
+         */
+        if (key == NULL || exact <= 0)
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+
+        /* Find a table to compact. */
+        while (ret == 0) {
+            WT_ERR(cursor->get_key(cursor, &key));
+            if (WT_PREFIX_MATCH(key, "file:") && !WT_STREQ(key, WT_HS_URI))
+                /* FIXME-WT-11343: check if the table is supposed to be compacted. */
+                break;
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+        }
+
+        /* We may have reached the end of the file. */
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+            full_iteration = true;
+            key = NULL;
+            continue;
+        }
+
+        /* Always close the metadata cursor. */
+        WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+
+        /* Compact the file. */
+        /* TODO: Make sure we are not reading a cached configuration. */
+        ret = wt_session->compact(wt_session, key, conn->background_compact.cfg);
```

Comment:
When should `background_compact.cfg` be updated? Should it be updated only when compact is called with the `background=true` configuration? Or every time session->compact is called? Including if the user is using "foreground compaction".

Reply:
Yes, it should be updated only when there a `WT_SESSION::compact` call with `background=true`.

Reply:
@Sean04, see this change: 0afb6625bad2a0b14ef7ba3b7353e18631b9f16a.
You can send a signal to the background compaction server using `__wt_compact_signal` which takes a configuration.